lib/component.js

```diff
@@ -255,6 +256,29 @@ class Component extends WebComponent {
     } else {
       this.el = this;
     }
+
+    const defaultContexts = this.getConfig(`defaultContexts`);
```

Comment:
so all Contexts have to be instantiated in the constructor, before any data is available to initialize them.

on the one hand, this saves you from the problem of "updating" a context and having to propagate that reference downwards.
on the other hand, once it's set, you can't update it or replace it, so maybe for more complicated contexts, the implementation just delegates behavior to some other object (like the component)

Reply:
If you had an example in mind, that would help in discussing the issues here.

I have thought about this, context replacements should not be "managed" by the `panel` framework at all, especially if the desired behavior is complex. Even if you ignore the implementation complications, at best it's a marginally useful helper method, while at worst you have to fight the framework or introduce more context framework concepts to control the context lifecycle to behave how you want. Complicated behavior is better off implemented in the specific scopes in which they are needed rather than centralized and fossilized into the framework.

Contexts shouldn't need to be "replaced" in most cases anyway. Contexts are designed to represent an environment, and you shouldn't need to "replace" your environment mid-way through. You also shouldn't need data to initialize a component context, much in the same way you don't need to "initialize" a browser with customer data: we tell the browser how to obtain data instead (by eg. giving it an API client, or a store knows how to query the API).